### ISSUE TITLE
Qt reactor fails two TCP tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11095]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1561|
|Type|defect|
|Created|2006-03-13 00:58:44Z|

Since [#7129](https://github.com/twisted/twisted/issues/7129) was merged, these have been problematic with Qt:

```
===============================================================================
[ERROR]: twisted.test.test_tcp.HalfCloseTestCase.testWriteCloseNotification

  File "/home/buildbot/BuildBot/slave/reactors/Twisted/twisted/test/test_tcp.py", line 1029, in testWriteCloseNotification
    f.protocol.transport.loseWriteConnection()
exceptions.AttributeError: 'NoneType' object has no attribute 'transport'
===============================================================================
[ERROR]: twisted.test.test_tcp.HalfCloseTestCase.testWriteCloseNotification

Failure: twisted.trial.unittest.FailTest: 0 != 1
-------------------------------------------------------------------------------
```

<details><summary>Searchable metadata</summary>

```
trac-id__1561 1561
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1142211524000000 1142211524000000
changetime__1147701666000000 1147701666000000
version__ 
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Done, ready for review in qtreactor-tcp-fixes-1561

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @exarkun|
|-|-|

Ok, looks like a pretty obvious fix (well, not that it was probably easy to figure out). 

Anyway, I think it's weird that you're assigning to client.closedDeferred instead of having client instantiate it. Would it hurt other tests to have that deferred created all the time?

NOTE: I DID NOT RUN THIS CODE ON QTREACTOR.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

In some ways, I'd prefer for the client to manage the Deferred lifetime itself.  However, this is a bit more complex than just having it assign the Deferred in __init__.  The code needs to be somewhat careful so that the Deferred is available when some other code wants to add callbacks to it, since at some point it disappears.  It'd be great to have a general refactoring of the entire testsuite to use a few carefully written classes with extremely awesome support for this kind of hook.

However, I think that's beyond the scope of this branch, so I've just done what most of the other tests do.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16363]) Merge qtreactor-tcp-fixes-1561

Author: exarkun
Reviewer: radix
Fixes [#11095](https://github.com/twisted/twisted/issues/11095)

This causes the tcp tests to wait on more events which they must wait on
in order to be correct.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

Milestone Twisted-2.3 deleted

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
